### PR TITLE
DAOS-3315 drpc: add prep shutdown plumbing

### PR DESCRIPTION
**Pull request by tanabarr**
Provide a prep_shutdown dRPC to be called as a result of gRPC fanout
to all members (daos_io_server processes) during the first phase of
controlled shutdown as described in DAOS-2883.

The dRPC handler (C side) will be used to trigger calls to disable
pool rebuild and auto eviction (on all members) as a precursor to
shutting down the DAOS system.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/server/mgmt_svc.go [hunk 1]
```diff
@@ -475,6 +475,40 @@ func (svc *mgmtSvc) SmdListPools(ctx context.Context, req *mgmtpb.SmdPoolReq) (*
...
+			mi = i
+			break
+		}
+	}
+
+	if mi == nil {
+		return nil, errors.Errorf("rank %d not found on this server", req.Rank)
+	}
+
+	dresp, err := mi.CallDrpc(drpc.ModuleMgmt, drpc.MethodPrepShutdown, req)
```
Comment (mjmac) [inline src/control/server/mgmt_svc.go]: Does it make sense to record that this operation was successful as an attribute on the membership entry? If I understand correctly, this method would be called via a compound operation prior to actually shutting down an instance, right? But if it fails for some reason, will that error get back to the admin? My concern is that we don't want the admin to request a shutdown which fails this step but then proceeds on to killing the instance, triggering a rebuild or other undesirable behavior.

Comment (tanabarr) [reply src/control/server/mgmt_svc.go]: yes this is part of the follow-up work that will perform the fanout call over gRPC, this PR is just a plumbing for the server side to put in place to handler where the evict disable and inhibit rebuild can be called from more details in https://jira.hpdd.intel.com/browse/DAOS-2883

Comment (tanabarr) [reply src/control/server/mgmt_svc.go]: I imagine we will want to store various state in the member records to enable decisions to be made based on condition of the members

File: src/mgmt/srv_drpc.c [hunk 2]
```diff
@@ -66,6 +66,43 @@ pack_daos_response(Mgmt__DaosResp *daos_resp, Drpc__Response *drpc_resp)
...
+ds_mgmt_drpc_prep_shutdown(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
+{
+	Mgmt__PrepShutdownReq	 *req = NULL;
+	Mgmt__DaosResp		 *resp = NULL;
+
+	/* Unpack the inner request from the drpc call body */
+	req = mgmt__prep_shutdown_req__unpack(
+		NULL, drpc_req->body.len, drpc_req->body.data);
+	if (req == NULL) {
+		drpc_resp->status = DRPC__STATUS__FAILURE;
```
Comment (kjacque) [inline src/mgmt/srv_drpc.c]: Just a note, you'll have to change this status code once #1544 lands.

File: src/mgmt/srv_drpc.c [hunk 3]
```diff
@@ -66,6 +66,43 @@ pack_daos_response(Mgmt__DaosResp *daos_resp, Drpc__Response *drpc_resp)
...
+		D_ERROR("Failed to unpack req (prep shutdown)\n");
+		return;
+	}
+
+	D_INFO("Received request to prep shutdown %u\n", req->rank);
+
+	D_ALLOC_PTR(resp);
```
Comment (kjacque) [inline src/mgmt/srv_drpc.c]: Suggestion - could make the Mgmt__DaosResp statically allocated instead. I've been taking that approach for many of the response structs, gives us fewer things to clean up.

File: src/mgmt/srv_drpc.c (hunk 2 above)
Comment (tanabarr) [reply src/mgmt/srv_drpc.c]: https://github.com/daos-stack/daos/pull/1591

File: src/mgmt/srv_drpc.c (hunk 3 above)
Comment (tanabarr) [reply src/mgmt/srv_drpc.c]: https://github.com/daos-stack/daos/pull/1591

**Pull request merged**